src/luarocks/manif/writer.tl

```diff
@@ -259,35 +253,41 @@ local function save_table(where, name, tbl)
    return ok, err
 end
 
-function writer.make_rock_manifest(name, version)
+function writer.make_rock_manifest(name: string, version: string): boolean, string
    local install_dir = path.install_dir(name, version)
-   local tree = {}
+   local tree: {string: Entry} = {}
    for _, file in ipairs(fs.find(install_dir)) do
       local full_path = dir.path(install_dir, file)
       local walk = tree
-      local last
-      local last_name
+      local last: {string : Entry}
+      local last_name: string
+      local next: Entry
       for filename in file:gmatch("[^\\/]+") do
-         local next = walk[filename]
+         next = walk[filename]
          if not next then
             next = {}
             walk[filename] = next
          end
          last = walk
          last_name = filename
-         walk = next
+         if not next is string then
+            walk = next
+         else
+            return nil, "Next was string, expected a table"
```

Comment:
this is an internal consistency check, so we shouldn't bubble-up an error message as if it was a normal failure condition.

If `next` is a string here, that means it was a LuaRocks bug. So, it's actually better to assert. Teal `next` is smart enough to remember an `is` test inside an assert:

```lua
last_name = filename
assert(next is {string: Entry})
walk = next
```